### DISCUSSION TITLE
Optimistic updates for a list

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2884] [Q&A] [answered]
I'm calling a `createTag` mutation, and I'd like the following outcome:

```ts
createTag({ tagName: '', organizationId })

// then I want to edit that organization
organization.tags = [optimisticTag, ...organizationTags]
```

What's the best way to do this? Is it with `updates.Mutation`? Or does that only run after the mutation has succeeded? I also find the API for `cache.updateQuery` confusing since I just want to update some fields on a keyed entity. I'd rather do something like `cache.invalidate`, but optimistically:

```tsx
cache.update({ __typename: 'Organization', id: organizationId }, data => {
  data.tags = [newTag, ...data.tags]

  return data
})
```

Is that sort of API possible? Even if it were, since it's inside of `updates.Mutation`, it's my understanding that this wouldn't even run optimistically.

Lastly, inside of `optimistic` resolvers, it seems that it only expects to get back what the mutation will return. In my case, `createTag` will return the newly-created `tag`. But how do I optimistically update `organization:id.tags`?

Thanks!

**Comment by JoviDeCroock** [comment, marked as answer]
Your `updates.Mutation` will run for optimistic results and for the actual results, this means that your logic you add for the actual result is usable for the optimistic result as well as they should have the same data-shape. When creating a tag you can return an optimistic result with a temp-id which will go through the same updater logic adding it to your list by means of `cache.updateQuery` or `cache.link`. When the actual result comes in we will clear the optimistic layer for you and apply the actual result (error or success).

For your case of adding to the tags of an organisation you could do:

```js
const entity = { __typename: 'Organization', id: organizationId }
const existingTags = cache.resolve(entity, 'tags')
cache.writeFragment(`
  fragment x on Organization { id tags }
`, { ...entity, tags: [...existingTags, newTag] })
```

or even

```js
const entity = { __typename: 'Organization', id: organizationId }
const existingTags = cache.resolve(entity, 'tags')
cache.link(entity, 'tags', [...existingTags, newTag])
```

You can also detect whether an updater is running for an optimistic update by means of the fourth `info` argument of the updater function having a property set named `optimistic` which will have `true` as a value.

**Reply by nandorojo** [reply]
Oh, wow, super useful info. I'm going to try all this out.

As for `cache.resolve(entity, 'tags')`, I don't see a type-safe way to use it. Would `cache.readFragment` have an equivalent implementation? If so, I'd likely prefer this since I can use `graphql()` from `graphql-codegen` inline with the `client-preset`.

This is what I'm trying now in `updates.Mutation`, based on your suggestion:

```ts
function createTag(parent, args, cache, info) {
  const __typename: Organization['__typename'] = 'Organization'
  const nestedTypename = 'tags' satisfies keyof Organization 

  const entity = {
    __typename,
    id: args.organizationId,
  }

  const tags: Organization[typeof nestedTypename] = cache.resolve(
    entity,
    nestedTypename
  ) || []

  cache.writeFragment(
    graphql(`
      fragment InsertTag_Cache on Organization {
        id
        tags {
          id
          name
          theme
          __typename
        }
      }
    `),
    {
      id: entity?.id,
      tags:[...tags, { __typename: 'Tag', id: `tag-${Date.now()}`, name: args.input.name, theme: args.input.theme }]
    }
  )
}
```

Here's an example of `cache.resolve`'s type unsafety:

<img width="895" alt="Screen Shot 2022-12-12 at 4 34 26 PM" src="https://user-images.githubusercontent.com/13172299/207159424-2e6c0178-9ab3-4d01-8f6f-ea813a340887.png">

**Reply by JoviDeCroock** [reply]
I mean, that's not really unsafe, you aren't getting back a `Tag[]` array you are getting back a list of strings which look like `Tag:<id>` just like they would look if you do `cache.keyOfEntity(entity)`, as shown [here](https://formidable.com/open-source/urql/docs/api/graphcache/#resolve). Anyway, that is probably how would I do it as well yeah.

**Reply by nandorojo** [reply]
Oh, I see. So the `existingTags` in your example is returning a list of tag IDs, not actual tag objects. I'm not exactly sure what the `newTag` should look like, then. Is that supposed to be the temporary ID I generated for my optimistic tag? Or does my implementation look correct?

**Reply by JoviDeCroock** [reply]
Both will work, when you return an object we will convert it into a key or you can do `cache.keyOfEntity(newTag)` but both should work just fine

**Reply by nandorojo** [reply]
I see, so this should work then. The only issue is that the type checker expects the `tags` to not be strings

<img width="1413" alt="Screen Shot 2022-12-12 at 4 40 23 PM" src="https://user-images.githubusercontent.com/13172299/207160400-bd3dc26b-dd42-406c-83b1-036906a3122d.png">
<img width="1260" alt="Screen Shot 2022-12-12 at 4 39 26 PM" src="https://user-images.githubusercontent.com/13172299/207160251-44cc96ae-2900-418a-bf6b-3ab904d3b89e.png">

Is that safe to ignore?

Also, do I have to add the `newTag` via `optimistic.createTag` too, or is just `updates.Mutation` sufficient? Thanks again.

**Reply by JoviDeCroock** [reply]
Optimistic serves as a way to "mock" what your API returns, the updater is an atomic way to update based on a result, so yes it's safe to only do it in an updater and just return the data in `optmistic`.

If you want to avoid the typing issue you can use `readFragment` and spread it in from there instead if you prefer :)

**Reply by nandorojo** [reply]
Got it, so to make sure I got this right:

**Step 1** `optimisic.createTag` creates some fake tag instantly

**Step 2** `updates.Mutation` is triggered by `optimistic.createTag`. For `updates.Mutation.createTag(parent)`, the `parent` argument is reading in what we returned from `Step 1`.

**Step 3** Once the network request resolves, `updates.Mutation` is called again, this time with real server's data, and the cache is updated once again. URQL knows the key of the optimistic result, so it deletes that item from the cache(?)

**Reply by JoviDeCroock** [reply]
The argument is the `result` 😅 in fact urql does not know about the key we use layering logic for clearing the optimistic result when a mutation completes, but again this is all [documented](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#optimistic-updates), also in the [intro](https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#deterministic-cache-updates).

Anything we can change about the docs to make this better?

**Reply by nandorojo** [reply]
> If you want to avoid the typing issue you can use readFragment and spread it in from there instead if you prefer :)

Good to know! I tried this, in a few places. However, it seems that `readFragment` wants me to define all fields, not just the ID.

For example, in my `optimistic.tagUser`, I want to assign an existing tag to a user. So I try to find it from the cache, and then return it optimistically:

```ts
tagUser(vars, cache, info) {
  const tag = cache.readFragment(
    graphql(`fragment ReadTag_Cache on Tag {
      id
      name
      theme
    }`), {
      id: vars.tagId,
    })

  return {
    __typename: 'Tag',
    id: `tag-${Date.now()}`, 
    name: tag?.name,
    theme: tag?.theme,
  }
}
```

My plan is to then apply this inside of `updates.Mutation`.

However, the types here for `readFragment` want more than just the `id`, requesting that I provide the `name` and `theme` as well, just to read it from the cache.

<img width="1327" alt="Screen Shot 2022-12-12 at 5 08 19 PM" src="https://user-images.githubusercontent.com/13172299/207165662-d1f5458b-075f-428c-ac70-6df2c9209642.png">

**Reply by nandorojo** [reply]
> Anything we can change about the docs to make this better?

Once I get this all figured out I'll put together some thoughts for making it clear!

I think having a full code sample in a single code block with both the optimistic update and `updates.Mutation`, and explaining when each gets called would help. I'll give it some thought.
